imap/jmap_mail.c

```diff
@@ -2835,19 +2835,49 @@ static int find_msgbodies(struct body *root, struct buf *msg_buf,
     return 0;
 }
 
+/* headers that are also used for non header.* properties are "special" */
+static int is_special_header (const char *key) {
+    if (strcasecmp(key, "sender") == 0) {
+        return 1;
+    }
+    else if (strcasecmp(key, "reply-to") == 0) {
+        return 1;
+    }
+    return 0;
+}
+
+struct extract_headers_rock {
+    json_t *headers;
+    hash_table *props;
+};
+
 static int extract_headers(const char *key, const char *val, void *rock)
 {
-    json_t *headers = (json_t*) rock;
+    json_t *headers = ((struct extract_headers_rock*) rock)->headers;
+    hash_table *props = ((struct extract_headers_rock*) rock)->props;
     json_t *curval;
     char *decodedval = NULL;
     char *lckey = xstrdup(key);
+    char *wantheader = NULL;
     char *p;
     for (p = lckey; *p; p++) {
         *p = tolower(*p);
     }
 
+    wantheader = strconcat("headers.", lckey, NULL);
+
+    if (!(   _wantprop(props, "headers")
+          || _wantprop(props, wantheader)
+          || is_special_header(lckey))) {
```

Comment:
Yep, but I don't really care for really complicated if statements, they're a nightmare to read/maintain.

I started out with a separate `is_special_header()` because in my head I expected there to be more of these cases... it was kind of a surprise to go through and find that it was only these two.  All the rest of the headers we care about here are cached in the body object -- so another option might be to cache these two headers in the body as well.  I don't know what the ramifications of that might be.

Now that I've seen how few headers are involved, I'm tempted to add the sender/reply-to mapping logic directly to `_wantprop()`.  That feels like a much better place for it.  What do you think?

Reply:
> another option might be to cache these two headers in the body as well

I agree, that would be cleanest solution. AFAIK it would blow up the index records, though, and we'd need to re-parse and re-write the records for all messages, if we don't want to special-case handle legacy records. Let's bring this up in today's Cyrus developer meeting, it would be good to understand what's the deal with index record updates.

For updating `_wantprop()`, I'm not sure I understand the proposed solution. Would be happy to align on this also in our meeting today.